nncf/experimental/torch/nas/bootstrapNAS/search/search.py

```diff
@@ -0,0 +1,386 @@
+        if ref_acc != 100:
+            self.search_params._ref_acc = ref_acc
+        self._tb = tensorboard_writer
+        self.checkpoint_save_dir = checkpoint_save_dir
+        if evaluators is not None:
+            self._use_default_evaluators = False
+            self._num_obj = len(evaluators)
+            self._evaluators = evaluators
+        else:
+            self._add_default_evaluators(validate_fn, val_loader)
+        self.update_evaluators_for_input_model()
+        self._problem = SearchProblem(self)
+        self._result = minimize(self._problem, self._algorithm,
+                                ('n_gen', int(self.search_params._num_evals / self.search_params._population)),
+                                seed=self.search_params._seed,
+                                # save_history=True,
+                                verbose=self._verbose)
+
+        if self._best_config is not None:
+            self._elasticity_ctrl.multi_elasticity_handler.set_config(self._best_config)
+            self._bn_adaptation.run(self._model)
+        else:
+            nncf_logger.warning("Couldn't find a subnet that satisfies the requirements. Returning maximum subnet.")
+            self._elasticity_ctrl.multi_elasticity_handler.activate_maximum_subnet()
+            self._bn_adaptation.run(self._model)
+            self._best_config = self._elasticity_ctrl.multi_elasticity_handler.get_active_config()
+            self._best_vals = [None, None]
+
+        return self._elasticity_ctrl, self._best_config, self._best_vals
+
+    def update_evaluators_for_input_model(self) -> NoReturn:
+        self._elasticity_ctrl.multi_elasticity_handler.activate_maximum_subnet()
+        for evaluator in self._evaluators:
+            if evaluator.type_of_measurement == 'accuracy':
+                value, _, _ = evaluator.evaluate_model(self._model)
+                evaluator.input_model_value = value
+                if value > self.search_params._ref_acc - 0.01 or value < self.search_params._ref_acc - 0.01:
+                    nncf_logger.warning(f"Accuracy obtained from evaluation {value} differs from reference accuracy {self.search_params._ref_acc}")
+                    if self.search_params._ref_acc == 100:
+                        nncf_logger.info("Adjusting reference accuracy to accuracy obtained from evaluation")
+                        self.search_params._ref_acc = value
+                    elif self.search_params._ref_acc < 100: # REMOVE. Not Needed because we want to make a distiction between both values.
+                        nncf_logger.info("Using reference accuracy.")
+                        evaluator.input_model_value = self.search_params._ref_acc
+            else:
+                if evaluator.use_model_for_evaluation:
+                    value = evaluator.evaluate_model(self._model)
+                else:
+                    value = evaluator.evaluate_with_elasticity_handler()
+                evaluator.input_model_value = value
```

Comment:
Here and everywhere, iteration over the list of evaluators inherited from a base class suggests usage of polymorphism. 
```
for evaluator in self._evaluators:
    evaluator.update_input_model_value(self._model)
```

`self.search_params._ref_acc` can be an internal field of evaluator, as well as other params.

Reply:
Fixed ref_acc in evaluator. Refactoring evaluators' loop.

Reply:
Refactored loop over evaluators and exposure of internals has been fixed. Thanks!